components/tikv_util/src/quota_limiter.rs

```diff
@@ -149,39 +196,98 @@ impl QuotaLimiter {
             .store(duration.0.as_nanos() as u64, Ordering::Relaxed);
     }
 
+    pub fn set_support_auto_tune(&self, support_auto_tune: bool) {
+        self.support_auto_tune
+            .store(support_auto_tune, Ordering::Relaxed);
+    }
+
+    pub fn foreground_cputime_limiter(&self) -> f64 {
+        self.foreground_cputime_limiter.speed_limit()
+    }
+
+    pub fn background_cputime_limiter(&self) -> f64 {
+        self.background_cputime_limiter.speed_limit()
+    }
+
     fn max_delay_duration(&self) -> Duration {
         Duration::from_nanos(self.max_delay_duration.load(Ordering::Relaxed))
     }
 
+    pub fn supports_auto_tune(&self) -> bool {
+        self.support_auto_tune.load(Ordering::Relaxed)
+    }
+
     // To generate a sampler.
     pub fn new_sample(&self) -> Sample {
         Sample {
             read_bytes: 0,
             write_bytes: 0,
             cpu_time: Duration::ZERO,
-            enable_cpu_limit: !self.cputime_limiter.speed_limit().is_infinite(),
+            enable_cpu_limit: !self.foreground_cputime_limiter.speed_limit().is_infinite()
+                || !self.background_cputime_limiter.speed_limit().is_infinite(),
+        }
+    }
+
+    // To consume a sampler and return delayed duration for foreground request.
+    // If the sampler is null, the speed limiter will just return ZERO.
+    pub async fn async_foreground_consume(&self, sample: Sample) -> Duration {
```

Comment:
how about extracting a inner limiter which limits cpu/write/read, and foreground and background can reuse the same logic